src/wallet/wallet.cpp

```diff
@@ -3024,82 +2982,161 @@ bool CWallet::SelectCoinsMinConf(const CAmount& nTargetValue, int nConfMine, int
+    std::vector<std::pair<CAmount, size_t>> amountsAndSizes = std::vector<std::pair<CAmount, size_t>>(vValue.size(), std::make_pair(0, 0));
+    for (int i = 0; i < vValue.size(); i++) {
+        amountsAndSizes[i] = std::make_pair(std::get<1>(vValue[i]), std::get<3>(vValue[i]));
+    }
+    
+    const CAmount targetAmount = nTargetValue;
+    const CAmount targetAmountPlusOffsetNoChange = nTargetValue; //no change
+    const CAmount targetAmountPlusOffsetMaxChange = std::get<2>(coinLowestLarger).tx != NULL ?
+                                                    std::min(std::get<1>(coinLowestLarger), nTotalLower) : nTotalLower; //max change
+    const size_t availableTotalSize = availableBytes;
+    std::unique_ptr<CCoinsSelectionAlgorithmBase> bestAlgorithm = nullptr;
+
+    std::unique_ptr<CCoinsSelectionAlgorithmBase> fastNotOptimalAlgorithm;
+    for (int i = 0; i < COINS_SELECTION_INTERMEDIATE_CHANGE_LEVELS + 2; ++i)
     {
-        setCoinsRet.insert(coinLowestLarger.second);
-        nValueRet += coinLowestLarger.first;
+        const CAmount targetAmountPlusOffset = targetAmountPlusOffsetNoChange +
+                                               (double)(i) / (COINS_SELECTION_INTERMEDIATE_CHANGE_LEVELS + 1) * (targetAmountPlusOffsetMaxChange - targetAmountPlusOffsetNoChange);
+        fastNotOptimalAlgorithm = std::unique_ptr<CCoinsSelectionAlgorithmBase>(new CCoinsSelectionSlidingWindow(amountsAndSizes, targetAmount, targetAmountPlusOffset, availableTotalSize));
+        // no async solving is required for these algorithms
+        fastNotOptimalAlgorithm->Solve();
+        if (fastNotOptimalAlgorithm->GetOptimalTotalSelection() > 0)
+        {
+            break;
+        }
     }
-    else {
-        for (unsigned int i = 0; i < vValue.size(); i++)
-            if (vfBest[i])
+
+    std::unique_ptr<CCoinsSelectionAlgorithmBase> slowOptimalAlgorithm = nullptr;
+    if (fastNotOptimalAlgorithm->GetOptimalTotalSelection() > 0)
+    {
+        slowOptimalAlgorithm = std::unique_ptr<CCoinsSelectionAlgorithmBase>(new CCoinsSelectionBranchAndBound(amountsAndSizes, targetAmount, fastNotOptimalAlgorithm->targetAmountPlusOffset, availableTotalSize));
+        // async solving is required for this algorithm
+        slowOptimalAlgorithm->StartSolvingAsync();
+        for (int wait = 0; wait < GetArg("-coinsselectiontimeout", 4) * 10; ++wait)
+        {
+            if (slowOptimalAlgorithm->GetHasCompleted())
             {
-                setCoinsRet.insert(vValue[i].second);
-                nValueRet += vValue[i].first;
+                break;
             }
+            std::this_thread::sleep_for(std::chrono::milliseconds(100));
+        }
+        if (!slowOptimalAlgorithm->GetHasCompleted())
+        {
+            slowOptimalAlgorithm->StopSolving();
+        }
+        CCoinsSelectionAlgorithmBase::GetBestAlgorithmBySolution(slowOptimalAlgorithm, fastNotOptimalAlgorithm, bestAlgorithm);
```

Comment:
In case `slowOptimalAlgorithm->GetHasCompleted()` returns false ... what is the reliability of its result (`optimalTotalSelection`) is constantly updated in the SolveRecursive routine ?

Reply:
Inside method `SolveRecursive` the value of `optimalTotalSelection` is constantly updated as soon as new leaves of the tree are explored. Hence, chances are that the incomplete solution (meaning a solution based on the leaves explored up to now) found by slowOptimalAlgorithm can be better than the complete one found by fastNotOptimalAlgorithm.

Reply:
I am pretty sure you understand how from a logical perspective it does not make any sense.
How a dev (maybe external) is supposed to understand an "aborted" call has produced a reliable result ?

Reply:
Added a comment in order to clarify the intent:
```
// the solution found by slowOptimalAlgorithm is checked even if the algorithm was unable to complete its full exploration
// due to timeout; since optimal solution is updated continuously, chances are that it is better than solution found by
// fastNotOptimalAlgorithm nevertheless
```